include/bw64/writer.hpp

```diff
@@ -73,22 +74,34 @@ namespace bw64 {
       writeChunk(dataChunk);
     }
 
-    /**
-     * @brief Finalize file
-     *
-     * This destructor will write all yet-to-be-written chunks to the file
-     * and will also finalize all required information, i.e. the final chunk
-     * sizes etc.
-     */
-    ~Bw64Writer() {
-      finalizeDataChunk();
-      for (auto chunk : postDataChunks_) {
-        writeChunk(chunk);
+    /// finalise and close the file
+    ///
+    /// Write all yet-to-be-written chunks to the file and finalize all
+    /// required information, i.e. the final chunk sizes etc.
+    ///
+    /// It is recommended to call this before the destructor, to handle
+    /// excpetions. If it does throw, this object may be in an invalid state,
+    /// so do not try again without creating a new object.
+    void close() {
+      if (!closed_) {
```

Comment:
Could this flag be factored out by always closing the file even if errors occurred? we're considering any call to this a one-shot thing anyway with no retry. So, e.g, wrap the finalise calls in a try/catch, and close on catch with a pass-through of the exception? Can just use if(fileStream_.is_open()) then for the condition. Perhaps bad practice to "try" though.

Reply:
I did briefly consider that, and went with this because i didn't really want to think about fstreams too much...

I think this is safe (supposedly `is_open` is only cleared by `close`), though maybe messier overall than one more flag. I've pushed a commit that changes this, what do you think?

Reply:
I slightly prefer not having the flag, but mainly because the var name didn't really reflect the state. I.e, it's not necessarily closed if closed_ is true - that flags really just being use to prevent repeated attempts to close. So I guess I was more thinking the var name either needs to reflect what's it's doing or made to really track closed state (in which case is_open does that anyway).
I'm unsure whether not closing might cause hanging open files too? it probably gets properly closed in any case when the object is destructed tbh.

Reply:
Yeah good point, i think I'll stick with the exception based one.

The "closed" naming was just to match the "close" function, which i think makes sense especially for things working with files.

> I'm unsure whether not closing might cause hanging open files too? it probably gets properly closed in any case when the object is destructed tbh.

I think it's fine, and works in testing, but I probably add a test of this in the normal set too (the `big` ones don't get ran in CI).

Reply:
Looking at it again, it seems odd to have `.close()` on writers but not readers, but it seems like a good opportunity to check for any file errors that haven't been spotted anyway, so i've added that. Sorry for the spam :)